src/anomalib/models/dsr/lightning_model.py

```diff
@@ -0,0 +1,184 @@
+            logger.info("Now training upsampling module.")
+
+    def training_step(self, batch: dict[str, str | Tensor]) -> STEP_OUTPUT:
+        """Training Step of DSR.
+
+        Feeds the original image and the simulated anomaly mask during first phase. During
+        second phase, feeds a generated anomalous image to train the upsampling module.
+
+        Args:
+            batch (dict[str, str | Tensor]): Batch containing image filename, image, label and mask
+
+        Returns:
+            STEP_OUTPUT: Loss dictionary
+        """
+        ph1_opt, ph2_opt = self.optimizers()
+
+        if self.current_epoch < self.second_phase:
+            # we are not yet training the upsampling module: we are only using the first optimizer
+            input_image = batch["image"]
+            # Create anomaly masks
+            anomaly_mask = self.quantized_anomaly_generator.augment_batch(input_image)
+            # Generate model prediction
+            model_outputs = self.model(input_image, anomaly_mask)
+            # Compute loss
+            loss = self.second_stage_loss(
+                model_outputs["recon_feat_hi"],
+                model_outputs["recon_feat_lo"],
+                model_outputs["embedding_bot"],
+                model_outputs["embedding_top"],
+                input_image,
+                model_outputs["obj_spec_image"],
+                model_outputs["pred_mask"],
+                model_outputs["true_mask"],
+            )
+
+            # compute manual optimizer step
+            ph1_opt.zero_grad()
+            self.manual_backward(loss)
+            ph1_opt.step()
+
+        else:
+            # we are training the upsampling module
+            input_image = batch["image"]
+            # Generate anomalies
+            input_image, anomaly_maps = self.perlin_generator.augment_batch(input_image)
+            # Get model prediction
+            model_outputs = self.model(input_image)
+            # Calculate loss
+            loss = self.third_stage_loss(model_outputs["pred_mask"], anomaly_maps)
+
+            # compute manual optimizer step
+            ph2_opt.zero_grad()
+            self.manual_backward(loss)
+            ph2_opt.step()
+
+        self.log("train_loss", loss.item(), on_epoch=True, prog_bar=True, logger=True)
+        return {"loss": loss}
+
+    def validation_step(self, batch: dict[str, str | Tensor], *args, **kwargs) -> STEP_OUTPUT:
+        """Validation step of DSR. The Softmax predictions of the anomalous class are used as anomaly map.
+
+        Args:
+            batch (dict[str, str | Tensor]): Batch of input images
+
+        Returns:
+            STEP_OUTPUT: Dictionary to which predicted anomaly maps have been added.
+        """
+        del args, kwargs  # These variables are not used.
+
+        model_outputs = self.model(batch["image"])
+        batch["anomaly_maps"] = model_outputs["pred_mask"]
+        batch["pred_scores"] = model_outputs["score"]
```

Comment:
Is it possible if `model_outputs` have the same keys for `anomaly_maps` and `pred_scores`? The reason for asking is that we need standard keys for the `TorchInferencer` stuff?

The tests are failing since the output model is returning a different output than what is expected here
https://github.com/openvinotoolkit/anomalib/blob/5eb20c64cc6ee8862bf2723b83fc4cae60a5594e/src/anomalib/deploy/inferencers/torch_inferencer.py#L155-L185

Reply:
I have create PR #1354 to handle cases like this. `DsrModel` needs to return `anomaly_map` and `pred_score` though.